### PR TITLE
[ADDED] Support for Websocket protocol

### PR DESCRIPTION
**Pull request by kozlovic**
Websocket protocol will be used as long as the URL scheme is `ws` or `wss`.
Note that no mixing of websocket and non-websocket URLs can be used (in the list of servers). Once a client connects, the gossip will be augmenting the URLs of that "type".

A `Compression()` option was added, which is currently documented as being used only for Websocket connections.

### CODE REVIEW

**Review by derekcollison — APPROVED** [review]
LGTM - Like the way you have organized things, will serve us well IMO.

Any data on performance changes?

**Comment by ripienaar** [comment]
Very exciting :)

**Comment by ripienaar** [comment]
LGTM, also curious about performance

**Comment by kozlovic** [comment]
Not sure if you are asking about performance of WS or impact of refactoring on non WS. For the later, quick dirty test.

From master:
```
$ go run examples/nats-bench/main.go -np 10 -ns 1 -n 10000000 foo
Starting benchmark [msgs=10000000, msgsize=128, pubs=10, subs=1]
NATS Pub/Sub stats: 5,511,378 msgs/sec ~ 672.78 MB/sec
 Pub stats: 2,755,819 msgs/sec ~ 336.40 MB/sec
..
 Sub stats: 2,755,838 msgs/sec ~ 336.41 MB/sec
```
From `ws` branch:
```
$ go run examples/nats-bench/main.go -np 10 -ns 1 -n 10000000 foo
Starting benchmark [msgs=10000000, msgsize=128, pubs=10, subs=1]
NATS Pub/Sub stats: 5,595,547 msgs/sec ~ 683.05 MB/sec
 Pub stats: 2,798,012 msgs/sec ~ 341.55 MB/sec
...
 Sub stats: 2,798,307 msgs/sec ~ 341.59 MB/sec
```
So really no impact on this test.
For Websocket perf:
```
$ go run examples/nats-bench/main.go -s ws://127.0.0.1:8080 -np 10 -ns 1 -n 10000000 foo
Starting benchmark [msgs=10000000, msgsize=128, pubs=10, subs=1]
NATS Pub/Sub stats: 4,921,103 msgs/sec ~ 600.72 MB/sec
 Pub stats: 2,460,663 msgs/sec ~ 300.37 MB/sec
...
 Sub stats: 2,460,640 msgs/sec ~ 300.37 MB/sec
```

**Comment by ripienaar** [comment]
Amazing.

File: ws.go [hunk 1]
```diff
@@ -0,0 +1,697 @@
...
+
+	wsFinalBit = 1 << 7
+	wsRsv1Bit  = 1 << 6 // Used for compression, from https://tools.ietf.org/html/rfc7692#section-6
+	wsRsv2Bit  = 1 << 5
+	wsRsv3Bit  = 1 << 4
+
+	wsMaskBit = 1 << 7
+
+	wsContinuationFrame     = 0
+	wsMaxFrameHeaderSize    = 14 // Since LeafNode may need to behave as a client
```
Comment (ColinSullivan1) [inline ws.go:54]: Just curious - will leafs use this?

**Comment by derekcollison** [comment]
That's awesome.

**Review by ColinSullivan1 — APPROVED** [review]
LGTM.  Nice!

File: ws.go (hunk 1 above)
Comment (kozlovic) [reply ws.go:54]: Sorry.. copied from server. Will remove, thanks!

**Comment by mirusky** [comment]
Hi guys, thanks for your work. 

Just some questions 

- When this will be available? 
- Or it's safe to use branch ws while this isn't merged?

**Comment by kozlovic** [comment]
@mirusky Thanks! We actually going to likely merge this in the main branch once the review is complete and Websocket will be part of the upcoming v1.11.0 release.

**Comment by kozlovic** [comment]
@ColinSullivan1 Please have a look at https://github.com/nats-io/nats.go/pull/719/commits/c8c0b60e25fd49179ec8a96daa8308fac1e9bef3 and see if it helps.

**Comment by ColinSullivan1** [comment]
LGTM - thanks Ivan!

**Pull request merged**